drivers/i2c/i2c_sf32lb.c

```diff
@@ -0,0 +1,306 @@
+	tcr |= I2C_TCR_TB;
+
+	if ((msg->len == 0) && i2c_is_stop_op(msg)) {
+		tcr |= I2C_TCR_STOP;
+	}
+
+	sys_write32(0, cfg->base + I2C_IER);
+	sys_clear_bits(cfg->base + I2C_CR, I2C_CR_LASTSTOP | I2C_CR_LASTNACK);
+	sys_clear_bits(cfg->base + I2C_SR, I2C_SR_TE | I2C_SR_BED);
+
+	sys_write8(addr, cfg->base + I2C_DBR);
+	sys_write32(tcr, cfg->base + I2C_TCR);
+
+	while (!sys_test_bit(cfg->base + I2C_SR, I2C_SR_TE_Pos)) {
+	}
+
+	sys_set_bit(cfg->base + I2C_SR, I2C_SR_TE_Pos);
+
+	if (sys_test_bit(cfg->base + I2C_SR, I2C_SR_NACK_Pos)) {
+		ret = -EIO;
+	}
+
+	return ret;
+}
+
+static int i2c_sf32lb_master_send(const struct device *dev, uint16_t addr, struct i2c_msg *msg)
+{
+	int ret = 0;
+	const struct i2c_sf32lb_config *cfg = dev->config;
+	uint32_t tcr = I2C_TCR_STOP;
+	uint32_t sr = 0;
+
+	ret = i2c_sf32lb_send_addr(dev, addr, msg);
+	if (ret < 0) {
+		return ret;
+	}
+
+	for (int j = 0; j < msg->len; j++) {
+		bool last = (msg->len - j == 1) ? true : false;
+
+		if (last) {
+			tcr |= I2C_TCR_STOP;
+		}
+
+		sys_write8(msg->buf[j], cfg->base + I2C_DBR);
+
+		sys_write32(tcr, cfg->base + I2C_TCR);
+
+		while (!sys_test_bit(cfg->base + I2C_SR, I2C_SR_TE_Pos)) {
+		}
+
+		sys_set_bit(cfg->base + I2C_SR, I2C_SR_TE_Pos);
+
+		if (sys_test_bit(cfg->base + I2C_SR, I2C_SR_NACK_Pos)) {
+			ret = -EIO;
+			break;
+		}
+	}
+	while (sys_test_bit(cfg->base + I2C_SR, I2C_SR_UB_Pos)) {
+	}
+
+	return ret;
+}
+
+static int i2c_sf32lb_master_recv(const struct device *dev, uint16_t addr, struct i2c_msg *msg)
+{
+	int ret = 0;
+	const struct i2c_sf32lb_config *cfg = dev->config;
+	uint32_t tcr = I2C_TCR_TB;
+
+	ret = i2c_sf32lb_send_addr(dev, addr, msg);
+	if (ret < 0) {
+		return ret;
+	}
+
+	for (int j = 0; j < msg->len; j++) {
+		bool last = (j == (msg->len - 1)) ? true : false;
+
+		if (last && i2c_is_stop_op(msg)) {
+			tcr |= (I2C_TCR_STOP | I2C_TCR_NACK);
+		}
+
+		sys_write32(tcr, cfg->base + I2C_TCR);
+
+		while (!sys_test_bit(cfg->base + I2C_SR, I2C_SR_RF_Pos)) {
+		}
+
+		sys_set_bit(cfg->base + I2C_SR, I2C_SR_RF_Pos);
+
+		msg->buf[j] = sys_read8(cfg->base + I2C_DBR);
+	}
+
+	while (sys_test_bit(cfg->base + I2C_SR, I2C_SR_UB_Pos)) {
+	}
+
+	return ret;
+}
+
+static int i2c_sf32lb_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_sf32lb_config *cfg = dev->config;
+	struct i2c_sf32lb_data *data = dev->data;
+	uint32_t cr = sys_read32(cfg->base + I2C_CR);
+	uint32_t dnf = FIELD_GET(I2C_CR_DNF_Msk, cr);
+	uint32_t clk_rate = sf32lb_clock_get_rate_dt(&config->clock);
+	uint32_t lv;
+	uint32_t slv;
+	uint32_t flv;
+
+	if (!(I2C_MODE_CONTROLLER & dev_config)) {
+		return -ENOTSUP;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		cr |= I2C_MODE_STD;
+		lv = (clk_rate / cfg->bitrate - dnf - 7 + 1) / 2;
+		slv = FIELD_PREP(I2C_LCR_SLV_Msk, lv);
+		sys_write32(slv, cfg->base + I2C_LCR);
+		break;
+
+	case I2C_SPEED_FAST:
+		cr |= I2C_MODE_FS;
+		lv = (clk_rate / cfg->bitrate - dnf - 7 + 1) / 2;
```

Comment:
formula should be documented

Reply:
also please, add clock control changes first, then squash latest commit so there are no I2C driver fixups